internal/controller/accessrequest_controller.go

```diff
@@ -33,30 +38,241 @@ type AccessRequestReconciler struct {
 	Scheme *runtime.Scheme
 }
 
+const (
+	// AccessRequestFinalizerName defines the name of the AccessRequest finalizer
+	// managed by this controller
+	AccessRequestFinalizerName = "accessrequest.ephemeral-access.argoproj-labs.io/finalizer"
+)
+
 // +kubebuilder:rbac:groups=ephemeral-access.argoproj-labs.io,resources=accessrequests,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=ephemeral-access.argoproj-labs.io,resources=accessrequests/status,verbs=get;update;patch
 // +kubebuilder:rbac:groups=ephemeral-access.argoproj-labs.io,resources=accessrequests/finalizers,verbs=update
 
 // Reconcile is part of the main kubernetes reconciliation loop which aims to
 // move the current state of the cluster closer to the desired state.
-// TODO(user): Modify the Reconcile function to compare the state specified by
-// the AccessRequest object against the actual cluster state, and then
-// perform operations to make the cluster state reflect the state specified by
-// the user.
 //
-// For more details, check Reconcile and its Result here:
-// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.18.2/pkg/reconcile
+// 1. verify if accessrequest is expired and status is "granted"
+// 1.1 if so, remove the user from the elevated role
+// 1.2 update the accessrequest status to "expired"
+// 1.3 return
+// 2. verify if user has the necessary access to be promoted
+// 2.1 if they don't, update the accessrequest status to "denied"
+// 2.2 return
+// 3. verify if CR is approved
```

Comment:
Probably need to add a step to validate the CR parameters such as "Is the requested time Duration <= MaxAllowedDuration"